### ISSUE TITLE
Refining of type parameters in the case of lists

### ISSUE THREAD

**Issue by camillol** [microsoft/pyright#7433]
I am pushing for Pyright adoption as the main type checker in a large codebase, and the single biggest stumbling block has been Pyright's inability to infer the element type of lists which are built starting from `[]`. This causes it to silently accept type errors that other type checkers catch (e.g. Mypy); more importantly, these are errors that users _expect_ a type checker to catch, and Pyright's behavior in this case keeps surprising people.

Similar issues have been brought up before; I have read through them, but I do not think the reporters gave a full account of the problem. I think it will be useful to investigate it through the lens of consistency with Pyright's behavior in similar situations, while addressing points previously raised.

### Programmer intent when building a list

Here is the basic problem case: building a list. This is an extremely common idiom in Python programs.

```python
def process_strings(x: list[str]) -> None: ...

def list_of_int() -> None:
    x = []                 # list[Unknown]
    for i in range(5):
        x.append(i)        # still list[Unknown]
    process_strings(x)     # no error because anything goes with Unknown
```

Pyright does not flag the call to `process_strings` as an error. That is surprising to users, since they have clearly built a list of `int`s and they are passing it to something that requires a list of `str`s.

In fact, if they use the _other_ common way of building a list of ints, Pyright _does_ detect the error:

```python
def list_of_int_comprehension() -> None:
    x = [i for i in range(5)]   # list[int]
    process_strings(x)          # pyright flags this, correctly
```

Programmer intent is equally clear in both approaches to building a list. In both cases a human developer would say that the code is building a list of ints. In both cases, the type is not explicitly mentioned, so the code _could_ do `x.append("hello")` later on. Pyright _could_ require writing `x: list[int] = [i for i in range(5)]` for the list comprehension, but it does not; instead, it recognizes that it is more useful to infer the intended type of the list from that of the elements.

The decision whether to use list comprehensions or `append` to build a list does not involve or indicate any difference in programmer intent; instead it is a matter of style and preference, and sometimes of necessity. The examples I am using here are simple and compact, but there are many circumstances in real code where using a list comprehension is not practical or even possible.

Therefore, I think it is reasonable to expect Pyright to infer the type of the list's elements in both cases. In practice, users _do_ expect this, and are surprised when it doesn't happen.

### More `append`s and list comprehensions

Suppose that we make the same inference for the loop as we do for the list comprehension, i.e. the append refines the type to `list[int]`. What should happen, then, if there is another `append` with a different type? By analogy, I think the inferred type should be broadened. For instance, these all mean the same thing:

```python
def more_appends() -> None:
    x = [str(i) if i % 2 else i for i in range(5)]
    # x is list[int|str]

    y = []
    for i in range(5):
        y.append(str(i) if i % 2 else i)
    # y should be list[int|str]

    z = []
    for i in range(5):
        if i % 2: 
            z.append(str(i))
        else:
            z.append(i)
    # z should be list[int|str]
```

Should we always allow `append` to broaden the type, if it has not been explicitly specified? Indeed we could, and there is precedent. We do broaden variable types by assignment:

```python
def broaden(b: bool) -> None:
    x = "hello"
    reveal_type(x)  # Literal['hello']
    if b:
        x = 2
    reveal_type(x)  # Literal['hello', 2]
```

Interestingly, the way Pyright currently handles modifying a list with an inferred type is actually inconsistent:

```python
def list_comprehension_then_append() -> None:
    y = [str(i) for i in range(5)]
    y.append(10)  # pyright flags an error here
    process_strings(y)

def list_comprehension_then_extend() -> None:
    y = [str(i) for i in range(5)]
    y.extend([10])  # pyright flags an error here
    process_strings(y)

def list_comprehension_plus() -> None:
    y = [str(i) for i in range(5)]
    y += [i for i in range(5)]  # this is accepted and turns y into list[int | str]
    process_strings(y)  # pyright flags an error here
```

It's not clear to me why `+=` and `extend` should be treated differently. They mean the same thing and modify the same object.

The main point I want to make here is that all the same questions about when to broaden occur when we infer the type of a list comprehension, and this was not considered an obstacle to making a useful inference for the common case. So we should not refuse to infer a useful type for `[]` + `append` until all possible ambiguities are eliminated.

Secondarily, we can argue that allowing `append` and `extend` to broaden an inferred type would improve consistency overall. It would be more consistent with the handling of inferred variable types, and it would resolve the inconsistency with `+=`.

However, even if we went with a more limited approach (e.g. "sealing" the inferred type after the first definition-level block that has refined it), that would still be a big improvement over the status quo.

### Inconsistent handling of `Unknown` arguments

Here is another inconsistency:

```python
def unknown_argument(b: bool, c) -> None:
    x = list(c)
    reveal_type(x)       # list[Unknown]
    process_strings(x)   # allowed
    if b:
        x = [2]
    reveal_type(x)       # list[int] | list[Unknown]
    process_strings(x)   # error

    y = list(c)
    reveal_type(y)       # list[Unknown]
    if b:
        y.append(2)
    reveal_type(y)       # list[Unknown]
    process_strings(y)   # allowed
```

The second `process_strings(x)` is an error because there is a code path where `x` contains an `int`, which is certainly invalid. But the same is true of `y`, and `process_strings(y)` should be treated as an error for the same reason. Here, refining the type to `list[Unknown | int]` would allow catching that error.

### `list[Unbound]`?

We should note, however, that `[]` should not simply be treated as `list[Unknown]` in the same sense as `list(c)` above. With an untyped argument, we truly have no idea what the type might be, so `y = list(c)` is a list that may contain anything; after `y.append(2)` we know that it certainly contains at least an `int`, but we have no guarantee that _all_ it contains are `int`s.

But if we see the list being initialized as empty and then having elements added to it (with the elements themselves having a known or inferred type, of course), then we know the type of all elements. Therefore, `[]` should have a different internal type; we could call it `list[Empty]`, but let's tentatively reuse `Unbound`. Then we could have this:

```python
    x = []           # list[Unbound]
    for i in range(5):
        x.append(i)
    reveal_type(x)   # should be list[int], not list[Unknown] nor list[Unknown | int]
```

### Review of previously raised issues

To show I have done my homework, and to respect the maintainers' time, let me briefly go through the points raised in previous related issues.

- https://github.com/microsoft/pyright/issues/5766
  - "This is the intended behavior for `list[Unknown]`"
    - A valid point. An untyped `[]` should instead be inferred as `list[Unbound]` or `list[Empty]` and the appropriate refining behavior should be implemented.
- https://github.com/microsoft/pyright/issues/6710 
  - "Narrowing would introduce false positives; intent unclear". Example:
    ```python
    # Today
    x = [] # Type of x is inferred to be `list[Unknown]`
    x.append(1) # No narrowing; x remains `list[Unknown]`
    x.append("a") # This produces no error

    # With change
    y = [] # Type of y is inferred to be `list[Unknown]`
    y.append(1) # Type of `y` is narrowed from `list[Unknown]` to `list[int]`
    y.append("a") # This would produce an error because "a" cannot be appended to a `list[int]`
    ```
    - See discussion of multiple `append`s above. In my opinion, this example pushes us towards simply allowing broadening of inferred types for subsequent `append`s. That would avoid false positives here, and it would be consistent with how we handle variables.
    - See also the discussion of intent in comparison with list comprehensions.
  - "It would requires hard-coding knoweldge of methods in `dict` and `list`"
    - It's not clear to me that it would. Typeshed tells us that `list` has a type parameter `_T` and that `append` takes a `_T`. If we treat `[]` as an list with an unbound `_T`, then `append` can narrow it. Subsequent `append`s can also broaden, just as assignments to the same variable in different paths can already broaden the inferred type.
    - In any case, `list` is such a fundamental building block of the language, and building a list by appending is such a common pattern, that special-casing it seems warranted if it turns out to be needed. If so, the special-casing should probably apply to all subtypes of `MutableSequence`.

### Conclusion

Building a list starting from `[]` is a basic Python idiom; failing to infer the list element type is inconsistent with Pyright's behavior with similar forms, and leaves Pyright unable to detect type issues that users would expect it to catch. In my experience trying to push for Pyright adoption in a large codebase, this is probably the largest stumbling block, as well as the foremost area in which Pyright lags behind other type checkers (e.g. Mypy, Pycharm).

I think it is possible to address this in a principled way that improves the internal consistency of Pyright's behavior, by treating it as inferring and refining an initially unbound type parameters, similarly to how the inferred type of variables gets refined.

**Comment by erictraut** [comment]
Thanks for the well-thought-out post.

I can understand why the behavior you're asking for would be desirable, but unfortunately what you're asking for would violate a bunch of design principles and rules that are used throughout pyright and the Python type system. For those reasons, it's unlikely that we would ever implement the behavior you're asking for here. I'm sure that's a disappointing and dissatisfying answer, so let me go into additional detail so you can better understand my perspective.

The application of type evaluation rules is largely dictated by the Python typing specification, and it's highly mechanistic in most cases. Type checkers are provided type information (through inlined type annotations or stub files), and they apply these rules to determine whether any constraints are violated. This mechanistic approach is generally a good thing because the rules are well-defined, predictable and understandable once you understand them. They're also designed to compose well with each other to produce consistent results. 

There are some aspects of the type system that are under-specified, but we're making efforts to close those holes and make Python type checkers more uniform and deterministic. As part of this effort, the typing community is working on type checker conformance tests. I've been investing significant time to make sure that pyright is the [most conformant](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) type checker available.

Most of the behaviors you're noting above result from the basic application of type rules and the type information provided in the [typeshed type stubs](https://github.com/python/typeshed) which are used by all major Python type checkers. For example, the `list` class is defined in the `builtins.pyi`, and it dictates how the `.append` or `.extend` method work for the `list` class. Pyright has no specialized knowledge of the `list` class and its methods; it simply applies the type information found in the type stub. You won't find any mention of the `append` or `extend` methods anywhere in the pyright code base, and that's generally a good thing.

Pyright _does_ have knowledge of Python language syntax and how the Python parser and interpreter treat them. These are not specified in type stubs, so they need to be hard-coded in type checkers. For example, pyright knows that `[x for x in y]` is syntax for a list comprehension and that this syntax will cause the Python interpreter to construct a new list at runtime.

As you're probably aware, many operators in Python have well-defined (but sometimes complex) runtime behaviors. For example, the `+=` operator invokes a piece of code that first calls the `__iadd__` on the LHS object. If that fails, it attempts to call the `__add__` method on the LHS object, and then it falls back to the `__radd__` method on the RHS object. Pyright effectively emulates this behavior at static analysis time. That explains why `y += [i for i in range(5)]` results in the type `list[str | int]`. The `list[int].__iadd__` method fails (i.e. it doesn't pass type checking), so pyright then attempts to use `list[int].__add__`, which succeeds (i.e. passes type checking) and produces the resulting type of `list[int | str]` according to the type stub definition.

Here are some specific principles use throughout pyright and the type system that would be violated if we were to implement the behaviors you're suggesting.
1. If a class or method is defined in typeshed, we use its definition as written. We don't override the behavior specified in the stubs. (In the past, we violated this principle in a couple of cases, and we ended up regretting it every time.)
2. The type of an object is established at the time it is constructed. A type checker should never change the type of an object from `list[int]` to a `list[int | str]` after it has been created. It is OK for the type of an expression (e.g. a variable name) to change if it is assigned a new value. It is also permissible to narrow the type of an expression (e.g. based on a conditional type guard), but the resulting narrowed type should be a subtype of the original type.
3. When calling a method on an object, the type of the object determines the (partially-specialized) type of the method. For example, if you call `append` on an object of type `list[str]`, the type of method `append` (which is initially `def append(self, _object: _T) -> None`) is bound to `list[str]` and partially specialized to `def append(__object: str) -> None`. The arguments passed to the call are then checked against the specialized type of the method. The type of a method _never_ influences the type of the object it is called on. That would be backward and would break a bunch of assumptions in the type system.

Static type checking is a powerful tool, but to take maximal advantage of it, you and your team will need to make small changes to your coding style. If you're too rigid and expect all of your existing coding approaches to "just work", you won't get the most value out of it.

In the case of list construction, if you are starting with an empty list (`[]` or `list()`), you should get into the habit of providing a type annotation for the variable that it's assigned to.

```python
def list_of_int() -> None:
    x: list[int] = []
    ...
```
